```{r}
library('dplyr', quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library('ggplot2', quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library('edgeR',quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("cowplot", quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library("GGally",quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library("DESeq2",quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library('readxl',quietly = TRUE,lib.loc="/usr/lib/R/site-library")
library("VennDiagram", quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("ComplexHeatmap",quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("flashClust", quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library('plotly' , quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library('tidyverse' ,quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library('htmlwidgets' , quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library('reshape2' ,quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("ggpubr" ,quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("car" ,quietly = TRUE, lib.loc="/usr/lib/R/site-library") 
library("goseq" ,quietly = TRUE, lib.loc="/usr/lib/R/site-library")
library("stringr",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("data.table",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("tidyr",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("dplyr",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("ggsignif",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("kableExtra",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("grid",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("gridExtra",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("bigmemory",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("doParallel",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("ggmanh",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("lme4",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("emmeans",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library("limma",quietly = TRUE , lib.loc="/usr/lib/R/site-library") 
library("MultiAssayExperiment",quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library('MOFA2', lib.loc="/usr/lib/R/site-library")
library('biomaRt', lib.loc="/usr/lib/R/site-library") # BioMart R package,  provides a convenient interface for accessing and using the Ensembl database.
library('clusterProfiler',quietly = TRUE , lib.loc="/usr/lib/R/site-library")
library('AnnotationDbi', quietly = TRUE ,lib.loc="/usr/lib/R/site-library")

```

#Obtain Ensembl Annotation


```{r}
#  BioMart object called cow connected to the Ensembl database for bovine genes

#  Retrieves these set of annotation data for all genes. 

#The Ensembl gene ID / the gene name / description of the gene / HGNC gene symbol /  gene biotype #(e.g., protein_coding, non_coding). / length of the transcript in base pairs / name of the #chromosome on which the gene is located  / start position of the gene on the chromosome / end #position of the gene on the chromosome.strand The strand of the chromosome on which the gene is #located (+ or -).


cow<-useMart("ensembl", 
             dataset = "btaurus_gene_ensembl",
             host="www.ensembl.org") 

annotation.ensembl.symbol<-getBM(attributes = c('ensembl_gene_id',
                                                'external_gene_name',
                                                'description','hgnc_symbol',
                                                'gene_biotype',
                                                'transcript_length',
                                                'chromosome_name',
                                                'start_position',
                                                'end_position','strand'), 
                                 values = "*",
                                 mart = cow)

#  Retrieves the same set of annotation data but only for protein-coding genes. 


annotation.ensembl.transcript<-getBM(attributes = c('ensembl_gene_id',
                                                    'gene_biotype',
                                                    'ensembl_transcript_id',
                                                    'transcript_length'), 
                                     values = "*", 
                                     mart = cow)

annotation.ensembl.transcript<-annotation.ensembl.transcript[annotation.ensembl.transcript$gene_biotype =="protein_coding",]

# Sorts gene ID in descending order by transcript length.
annotation.ensembl.transcript<-annotation.ensembl.transcript[order(annotation.ensembl.transcript$ensembl_gene_id, -annotation.ensembl.transcript$transcript_length),]


# Removes any duplicate gene IDs
annotation.ensembl.transcript<-annotation.ensembl.transcript[!duplicated(annotation.ensembl.transcript$ensembl_gene_id),]

# Include transcripts that are at least 400 base pairs long.
annotation.ensembl.transcript<-annotation.ensembl.transcript[annotation.ensembl.transcript$transcript_length > 400,]

annotation.ensembl.symbol<-annotation.ensembl.symbol[order(annotation.ensembl.symbol$ensembl_gene_id, -annotation.ensembl.symbol$transcript_length),]

annotation.ensembl.symbol<-annotation.ensembl.symbol[!duplicated(annotation.ensembl.symbol$ensembl_gene_id),]

gene.length<-annotation.ensembl.symbol[,c( "ensembl_gene_id", "transcript_length" )]

annotation.GO.biomart<-getBM(attributes = c('ensembl_gene_id',
                                            'external_gene_name',
                                            'go_id',
                                            'name_1006',
                                            'namespace_1003'),
                             values = "*",
                             mart = cow)
```
```{r}
write.table(annotation.ensembl.symbol,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.symbol.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)
#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.symbol.txt")

write.table(gene.length,
            file="/home/yassine/Downloads/2023_10_10_btaurus_gene.length.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_gene.length.txt")

write.table(annotation.GO.biomart,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.GO.biomart.txt", 
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.GO.biomart.txt")

write.table(annotation.ensembl.transcript,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.transcript.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.transcript.txt")
```


# Read the count data
```{r}
files<-list.files("/mnt/storage/lab_folder/repro_omics/counting/", 
                  recursive=T,
                  pattern="count", 
                  full.names = TRUE)

files<-files[grep("summary", files, invert = TRUE)]
files<-files[grep(".count1", files, invert = TRUE)]
files<-files[grep(".count2", files, invert = TRUE)]
files<-files[grep(".sh", files, invert = TRUE)]
```

```{r}
count_data<-data.frame(matrix(nrow=27607))
for (n in 1:132) {
  count<-read.delim(files[n],
                    header=TRUE, 
                    sep= "\t", 
                    stringsAsFactors = FALSE,
                    comment.char= "#")
  
  count<-count[,c(1,7)]
  count_data<-cbind(count_data,count)
}

rownames(count_data)<-count_data[,2]
count_data<-count_data[,seq(from = 3, 
                            to = 265,
                            by = 2)]
colnames(count_data)<- substr(colnames(count_data),57, 59)
```


# Import Sample Information

```{r}
sample_info <- as.data.frame(read_excel("/home/yassine/Downloads/Repro_omics_sample_information_06_27_2023.xlsx"))
```

# Wrangling and data exploration

```{r}
sample_info_ <- sample_info[, -c(4,5)]

sample_info_ <- sample_info_[!is.na(sample_info_$heiferID), ]

sample_info_ <- sample_info_ %>%
  mutate(month = as.integer(gsub("-month", "", month))) %>%
  dplyr::rename(tissue = sample)


```
# Counts of unique months and unique tissues per HeiferID

```{r}
heifer_summary <- sample_info_ %>%
  group_by(heiferID) %>%
  summarise(unique_months = n_distinct(month),
            unique_tissues = n_distinct(tissue))
print(heifer_summary)

```

# Create a summary dataframe with counts of unique tissues and their names per month

```{r}
month_tissue_summary <- sample_info_ %>%
  group_by(month) %>%
  summarise(
    distinct_tissues = n_distinct(tissue),
    tissue_names = paste(unique(tissue), collapse = ", ")
  )
print(month_tissue_summary)

```

```{r}
# Create a vector of tissue names for the 19 months samples
tissue_19_months <- sample_info_ %>%
  filter(month == 19) %>%
  pull(tissue)

# Create a vector of tissue names for the 6 months samples
tissue_6_months <- sample_info_ %>%
  filter(month == 6) %>%
  pull(tissue)

# Create a vector of tissue names for the 9 months samples
tissue_9_months <- sample_info_ %>%
  filter(month == 9) %>%
  pull(tissue)

# Tissues present in 19 months but not in 6 months
tissues_in_19_not_in_6 <- setdiff(tissue_19_months, tissue_6_months)

# Tissues present in 19 months but not in 9 months
tissues_in_19_not_in_9 <- setdiff(tissue_19_months, tissue_9_months)

# Tissues present in 6 months but not in 9 months
tissues_in_6_not_in_9 <- setdiff(tissue_6_months, tissue_9_months)

# Print the results
cat("Tissues present in 19 months but not in 6 months:", paste(tissues_in_19_not_in_6, collapse = ", "), "\n")
cat("Tissues present in 19 months but not in 9 months:", paste(tissues_in_19_not_in_9, collapse = ", "), "\n")
cat("Tissues present in 6 months but not in 9 months:", paste(tissues_in_6_not_in_9, collapse = ", "), "\n")

```

```{r}
# Get a vector of unique tissues
unique_tissues <- unique(sample_info_$tissue)

  # Specify the target month
for (target_month in c(6,9,19)){
  
  # Loop through unique tissues and print distinct samples for each
  for (tissue in unique_tissues) {
    filtered_data <- sample_info_ %>%
      filter(month == target_month, tissue == tissue)
    
    distinct_samples <- unique(filtered_data$heiferID)
    
    cat("For Month:", target_month, "and Tissue:", tissue, "\n")
    cat("Distinct Samples (heifer_IDs):", "\n",
        "number:" , length(distinct_samples), "\n",
        "IDs", paste(distinct_samples, collapse = ", "), "\n")
  }
}
```

```{r}
sample_info_replicas <- sample_info_ %>%
  group_by(heiferID, month, tissue) %>%
  mutate(replica_count = n()) %>%
  filter(replica_count > 1) %>%
  distinct(heiferID, month, tissue, .keep_all = TRUE)

# Print the replica entries with their replica counts
print(sample_info_replicas)
```
# creation of the data_info for the co-expression analysis

```{r}
data_info <- sample_info[, -4] %>% # Remove the fourth column
  filter(!is.na(sample_info[[5]])) %>% #Keep rows with non-NA entries in the fifth column (tissu)
  dplyr::rename(tissu = sample)%>% # Rename the 'sample' column to 'tissu'
  filter(!is.na(heiferID))%>% #Remove rows with NA entry in the 'heiferID' column
  mutate(month = as.integer(gsub("-month", "", month)))#Convert character entries in the 'month' column to integers

colnames(data_info)[4] <- "sample"
```


# Summary dataframe for all samples (counts of unique tissues and their names per month)

```{r}
month_tissue_summary_ <- data_info %>%
  group_by(month) %>%
  summarise(
    tissues = paste(unique(tissu), collapse = ", "),
    distinct_samples = n_distinct(heiferID),
    Heifer_IDs = paste(unique(heiferID), collapse = ", ")
  )

# Print the modified summary dataframe
print(month_tissue_summary_)
```

```{r}
data_info$sample<- as.factor(data_info$sample)
data_info$month<- as.factor(data_info$month)
data_info$tissu<- as.factor(data_info$tissu)
data_info$heiferID<- as.factor(data_info$heiferID)
colnames(count_data)<-data_info$heiferID

count_data_export<-data.frame(gene_id=rownames(count_data),count_data)
```


```{r}
write_delim(count_data_export,
            file = "/home/yassine/Downloads/count_data.txt", 
            delim = "\t",
            quote =  "none")

#system("bzip2 "/home/yassine/Downloads/count_data.txt")

```

#Sequencing summary


```{r}
count_data_a<-count_data[rowSums(count_data)>0,] #filtering genes that have no reads 

# Adds the annotation from "btaurus_gene_ensembl" to the filtred gene counts  
count_data_annotated<-merge(count_data_a,
                            annotation.ensembl.symbol,
                            by.x="row.names",
                            by.y="ensembl_gene_id",
                            all.x=TRUE,
                            all.y=FALSE)
```

```{r}
data_info <- sample_info[, -4] %>% # Remove the fourth column
  filter(!is.na(sample_info[[5]])) %>% #Keep rows with non-NA entries in the fifth column (tissu)
  dplyr::rename(tissu = sample)%>% # Rename the 'sample' column to 'tissu'
  filter(!is.na(heiferID))%>% #Remove rows with NA entry in the 'heiferID' column
  mutate(month = as.integer(gsub("-month", "", month)))#Convert character entries in the 'month' column to integers

colnames(data_info)[4] <- "sample"
data_info$sample_name <- paste0(data_info$heiferID, "_", data_info$tissu,"_", data_info$month) #To avoid having duplicate samples later 
```

```{r}
data_info$sample<- as.factor(data_info$sample)
data_info$month<- as.factor(data_info$month)
data_info$tissu<- as.factor(data_info$tissu)
data_info$heiferID<- as.factor(data_info$heiferID)
```

```{r}
write.table(data_info,
            file = "/home/yassine/Downloads/between_tissue/data_info.txt",
            quote = TRUE,
            sep = "\t",
            row.names = TRUE,
            col.names = TRUE)
```


```{r}
colnames(count_data_a)<-data_info$sample_name

count_data_a<-data.frame(gene_id=rownames(count_data_a),count_data_a)

write_delim(count_data_a,
            file = "/home/yassine/Downloads/count_data_a.txt", 
            delim = "\t",
            quote =  "none")
```


```{r}
# Count the occurrences of each gene biotype
biotype_counts <- table(count_data_annotated$gene_biotype)

# Extract the counts for protein_coding, lncRNA, pseudogenes, and others
protein_coding_count <- biotype_counts["protein_coding"]
lncrna_count <- biotype_counts["lncRNA"]
pseudogenes_count <- biotype_counts["pseudogene"]
others_count <- nrow(count_data_annotated) - protein_coding_count - lncrna_count - pseudogenes_count

# Print the counts
print(paste("protein_coding:", protein_coding_count))
print(paste("lncRNA:", lncrna_count))
print(paste("pseudogenes:", pseudogenes_count))
print(paste("others:", others_count))
```

```{r}
# Counts the number of protein coding genes retreived from RNA_seq reads 
n_reads_protein_coding<-sum(count_data_annotated[count_data_annotated$gene_biotype=='protein_coding',c(2:6)])

# Counts the number of lon non coding RNA genes retreived from RNA_seq reads 
n_reads_lncRNA<-sum(count_data_annotated[count_data_annotated$gene_biotype=='lncRNA',c(2:6)])

# Counts the number of pseudogenes retreived from RNA_seq reads 
n_reads_pseudogene<-sum(count_data_annotated[count_data_annotated$gene_biotype %in% c('pseudogene','processed_pseudogene'),c(2:6)])

# Counts the number of reads that doesn't belong to any of the categories above r
n_reads_others<-sum(count_data_annotated[!(count_data_annotated$gene_biotype %in% c('pseudogene','processed_pseudogene','protein_coding','lncRNA')),c(2:6)])

summary_RNA_seq<-data.frame(class=c('protein_coding',
                                    'lncRNA', 
                                    'pseudogene',
                                    'others'),
                            nreads=c(n_reads_protein_coding,
                                     n_reads_lncRNA,
                                     n_reads_pseudogene,
                                     n_reads_others))

summary_RNA_seq<-mutate(summary_RNA_seq,
                        prop = round(nreads/sum(nreads)*100,2))

summary_RNA_seq<-summary_RNA_seq[with(summary_RNA_seq, order(-nreads)),]

summary_RNA_seq$class<-factor(summary_RNA_seq$class,
                              levels=c('protein_coding', 'lncRNA','pseudogene','others' ))
```

```{r}
knitr::kable(summary_RNA_seq,
             format = 'pandoc')
```

```{r}
# Count_data_annotated_length contains the read counts and the transcript_lengths for the genes that generates one of these RNA types 'protein_coding', 'lncRNA','pseudogene' the count_data_annotated data frame.

count_data_annotated<-count_data_annotated[count_data_annotated$gene_biotype %in% c('protein_coding', 'lncRNA','pseudogene'),]


```

```{r}
count_data_b<-count_data_annotated[,c(2:133)]
rownames(count_data_b)<-count_data_annotated$Row.names
colnames(count_data_b)<-data_info$sample_name

```

```{r}
write_delim(count_data_b,
            file = "/home/yassine/Downloads/count_data_b.txt", 
            delim = "\t",
            quote =  "none")
```
