#Obtain Ensembl Annotation


```{r}
#  BioMart object called cow connected to the Ensembl database for bovine genes

#  Retrieves these set of annotation data for all genes. 

#The Ensembl gene ID / the gene name / description of the gene / HGNC gene symbol /  gene biotype #(e.g., protein_coding, non_coding). / length of the transcript in base pairs / name of the #chromosome on which the gene is located  / start position of the gene on the chromosome / end #position of the gene on the chromosome.strand The strand of the chromosome on which the gene is #located (+ or -).


cow<-useMart("ensembl", 
             dataset = "btaurus_gene_ensembl",
             host="www.ensembl.org") 

annotation.ensembl.symbol<-getBM(attributes = c('ensembl_gene_id',
                                                'external_gene_name',
                                                'description','hgnc_symbol',
                                                'gene_biotype',
                                                'transcript_length',
                                                'chromosome_name',
                                                'start_position',
                                                'end_position','strand'), 
                                 values = "*",
                                 mart = cow)

#  Retrieves the same set of annotation data but only for protein-coding genes. 


annotation.ensembl.transcript<-getBM(attributes = c('ensembl_gene_id',
                                                    'gene_biotype',
                                                    'ensembl_transcript_id',
                                                    'transcript_length'), 
                                     values = "*", 
                                     mart = cow)

annotation.ensembl.transcript<-annotation.ensembl.transcript[annotation.ensembl.transcript$gene_biotype =="protein_coding",]

# Sorts gene ID in descending order by transcript length.
annotation.ensembl.transcript<-annotation.ensembl.transcript[order(annotation.ensembl.transcript$ensembl_gene_id, -annotation.ensembl.transcript$transcript_length),]


# Removes any duplicate gene IDs
annotation.ensembl.transcript<-annotation.ensembl.transcript[!duplicated(annotation.ensembl.transcript$ensembl_gene_id),]

# Include transcripts that are at least 400 base pairs long.
annotation.ensembl.transcript<-annotation.ensembl.transcript[annotation.ensembl.transcript$transcript_length > 400,]

annotation.ensembl.symbol<-annotation.ensembl.symbol[order(annotation.ensembl.symbol$ensembl_gene_id, -annotation.ensembl.symbol$transcript_length),]

annotation.ensembl.symbol<-annotation.ensembl.symbol[!duplicated(annotation.ensembl.symbol$ensembl_gene_id),]

gene.length<-annotation.ensembl.symbol[,c( "ensembl_gene_id", "transcript_length" )]

annotation.GO.biomart<-getBM(attributes = c('ensembl_gene_id',
                                            'external_gene_name',
                                            'go_id',
                                            'name_1006',
                                            'namespace_1003'),
                             values = "*",
                             mart = cow)
```
```{r}
write.table(annotation.ensembl.symbol,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.symbol.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)
#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.symbol.txt")

write.table(gene.length,
            file="/home/yassine/Downloads/2023_10_10_btaurus_gene.length.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_gene.length.txt")

write.table(annotation.GO.biomart,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.GO.biomart.txt", 
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.GO.biomart.txt")

write.table(annotation.ensembl.transcript,
            file="/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.transcript.txt",
            sep = "\t",
            append = FALSE,
            quote = FALSE)

#system('bzip2 --best "/home/yassine/Downloads/2023_10_10_btaurus_annotation.ensembl.transcript.txt")
```


# Read the count data
```{r}
files<-list.files("/mnt/storage/lab_folder/repro_omics/counting/", 
                  recursive=T,
                  pattern="count", 
                  full.names = TRUE)

files<-files[grep("summary", files, invert = TRUE)]
files<-files[grep(".count1", files, invert = TRUE)]
files<-files[grep(".count2", files, invert = TRUE)]
files<-files[grep(".sh", files, invert = TRUE)]
```

```{r}
count_data<-data.frame(matrix(nrow=27607))
for (n in 1:132) {
  count<-read.delim(files[n],
                    header=TRUE, 
                    sep= "\t", 
                    stringsAsFactors = FALSE,
                    comment.char= "#")
  
  count<-count[,c(1,7)]
  count_data<-cbind(count_data,count)
}

rownames(count_data)<-count_data[,2]
count_data<-count_data[,seq(from = 3, 
                            to = 265,
                            by = 2)]
colnames(count_data)<- substr(colnames(count_data),57, 59)
```


# Import Sample Information

```{r}
sample_info <- as.data.frame(read_excel("/home/yassine/Downloads/Repro_omics_sample_information_06_27_2023.xlsx"))
```

# Wrangling and data exploration

```{r}
sample_info_ <- sample_info[, -c(4,5)]

sample_info_ <- sample_info_[!is.na(sample_info_$heiferID), ]

sample_info_ <- sample_info_ %>%
  mutate(month = as.integer(gsub("-month", "", month))) %>%
  dplyr::rename(tissue = sample)


```
# Counts of unique months and unique tissues per HeiferID

```{r}
heifer_summary <- sample_info_ %>%
  group_by(heiferID) %>%
  summarise(unique_months = n_distinct(month),
            unique_tissues = n_distinct(tissue))
print(heifer_summary)

```
